panic on gopkg.in .v0 imports

### What version of Go (`go version`) and `dep` (`git describe --tags`) are you using?
```
[I] 13:21:22 ~/D/g/s/g/d/testboom $ go version
go version go1.8.3 darwin/amd64

[I] 13:23:56 ~/D/g/s/g/g/dep (master) $ git describe --tags
v0.1.0-313-g44a454b
```

### What `dep` command did you run?

`dep init -v`

```
Root project is "github.com/daenney/testboom"
 1 transitively valid internal packages
 1 external packages imported from 1 projects
(0)   ✓ select (root)
(1)	? attempt gopkg.in/go-ldap/ldif.v0 with 1 pkgs; 0 versions to try
panic: version queue is empty, should not happen

goroutine 1 [running]:
github.com/golang/dep/internal/gps.(*solver).findValidVersion(0xc42016eb00, 0xc4201de1e0, 0xc420153fe0, 0x1, 0x1, 0x0, 0xc420153fe0)
	/Users/daenney/Development/go/src/github.com/golang/dep/internal/gps/solver.go:845 +0x5fa
github.com/golang/dep/internal/gps.(*solver).createVersionQueue(0xc42016eb00, 0xc420180b01, 0x18, 0x0, 0x0, 0xc420153fe0, 0x1, 0x1, 0x0, 0x0, ...)
	/Users/daenney/Development/go/src/github.com/golang/dep/internal/gps/solver.go:832 +0xace
github.com/golang/dep/internal/gps.(*solver).solve(0xc42016eb00, 0x0, 0x0, 0xc42019a0f0)
	/Users/daenney/Development/go/src/github.com/golang/dep/internal/gps/solver.go:449 +0x471
github.com/golang/dep/internal/gps.(*solver).Solve(0xc42016eb00, 0x3d, 0x16ce440, 0xc42019a030, 0xc4200140c6)
	/Users/daenney/Development/go/src/github.com/golang/dep/internal/gps/solver.go:399 +0xb1
main.(*initCommand).Run(0xc420153a6a, 0xc420015950, 0xc420010470, 0x0, 0x0, 0x0, 0x0)
	/Users/daenney/Development/go/src/github.com/golang/dep/cmd/dep/init.go:172 +0x710
main.(*Config).Run(0xc42001b080, 0xc42001b080)
	/Users/daenney/Development/go/src/github.com/golang/dep/cmd/dep/main.go:158 +0x876
main.main()
	/Users/daenney/Development/go/src/github.com/golang/dep/cmd/dep/main.go:45 +0x253
```

### What did you expect to see?

An error informing me that `.v0` URLs aren't really useful and just use https://github.com/go-ldap/ldif directly or it working as expected.

### What did you see instead?

A panic.

### Reproduce:

```
[I] 13:24:17 ~/D/g/s/g/d/testboom $ pwd
/Users/daenney/Development/go/src/github.com/daenney/testboom

[I] 13:25:24 ~/D/g/s/g/d/testboom $ cat main.go
package main

import (
	"gopkg.in/go-ldap/ldif.v0"
)

func main() {
	l := &ldif.LDIF{}
	print(l)
}

[I] 13:25:36 ~/D/g/s/g/d/testboom $ dep init -v
Root project is "github.com/daenney/testboom"
...
```

Doesn't look like it:

```
Root project is "github.com/daenney/testboom"
 1 transitively valid internal packages
 1 external packages imported from 1 projects
(0)   ✓ select (root)
(1)	? attempt gopkg.in/go-ldap/ldif.v0 with 1 pkgs; 0 versions to try
panic: version queue is empty, should not happen

goroutine 1 [running]:
github.com/golang/dep/internal/gps.(*solver).findValidVersion(0xc420144400, 0xc4202360a0, 0xc420045c90, 0x1, 0x1, 0x0, 0xc420045c90)
	/Users/daenney/Development/go/src/github.com/golang/dep/internal/gps/solver.go:845 +0x589
github.com/golang/dep/internal/gps.(*solver).createVersionQueue(0xc420144400, 0xc420162dc1, 0x18, 0x0, 0x0, 0xc420045c90, 0x1, 0x1, 0x0, 0x0, ...)
	/Users/daenney/Development/go/src/github.com/golang/dep/internal/gps/solver.go:832 +0xaa2
github.com/golang/dep/internal/gps.(*solver).solve(0xc420144400, 0x0, 0x0, 0xc420180f30)
	/Users/daenney/Development/go/src/github.com/golang/dep/internal/gps/solver.go:449 +0x4cf
github.com/golang/dep/internal/gps.(*solver).Solve(0xc420144400, 0x3d, 0x16ff640, 0xc420180e10, 0xc4200140c6)
	/Users/daenney/Development/go/src/github.com/golang/dep/internal/gps/solver.go:399 +0xaa
main.(*initCommand).Run(0xc420155802, 0xc420086f50, 0xc4200101d0, 0x0, 0x0, 0x0, 0x0)
	/Users/daenney/Development/go/src/github.com/golang/dep/cmd/dep/init.go:172 +0x6ef
main.(*Config).Run(0xc420062f00, 0xc420062f00)
	/Users/daenney/Development/go/src/github.com/golang/dep/cmd/dep/main.go:158 +0x768
main.main()
	/Users/daenney/Development/go/src/github.com/golang/dep/cmd/dep/main.go:45 +0x1fc
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,7 @@ BUG FIXES:
 * Releases targeting Windows now have a `.exe` suffix (#1291).
 * Adaptively recover from dirty and corrupted git repositories in cache (#1279).
 * Suppress git password prompts in more places (#1357).
+* Support [gopkg.in version zero](http://labix.org/gopkg.in#VersionZero). Fixes potential panic when referencing v0 gopkg.in packages  ([#1243](https://github.com/golang/dep/pull/1243)).
 
 IMPROVEMENTS:
 

--- a/gps/vcs_source.go
+++ b/gps/vcs_source.go
@@ -384,23 +384,34 @@ func (s *gopkginSource) listVersions(ctx context.Context) ([]PairedVersion, erro
 	k := 0
 	var dbranch int // index of branch to be marked default
 	var bsv semver.Version
+	var defaultBranch PairedVersion
+	tryDefaultAsV0 := s.major == 0
 	for _, v := range ovlist {
 		// all git versions will always be paired
 		pv := v.(versionPair)
 		switch tv := pv.v.(type) {
 		case semVersion:
+			tryDefaultAsV0 = false
 			if tv.sv.Major() == s.major && !s.unstable {
 				vlist[k] = v
 				k++
 			}
 		case branchVersion:
+			if tv.isDefault && defaultBranch == nil {
+				defaultBranch = pv
+			}
+
 			// The semver lib isn't exactly the same as gopkg.in's logic, but
 			// it's close enough that it's probably fine to use. We can be more
 			// exact if real problems crop up.
 			sv, err := semver.NewVersion(tv.name)
-			if err != nil || sv.Major() != s.major {
-				// not a semver-shaped branch name at all, or not the same major
-				// version as specified in the import path constraint
+			if err != nil {
+				continue
+			}
+			tryDefaultAsV0 = false
+
+			if sv.Major() != s.major {
+				// not the same major version as specified in the import path constraint
 				continue
 			}
 
@@ -433,6 +444,12 @@ func (s *gopkginSource) listVersions(ctx context.Context) ([]PairedVersion, erro
 			name:      dbv.v.(branchVersion).name,
 			isDefault: true,
 		}.Pair(dbv.r)
+	}
+
+	// Treat the default branch as v0 only when no other semver branches/tags exist
+	// See http://labix.org/gopkg.in#VersionZero
+	if tryDefaultAsV0 && defaultBranch != nil {
+		vlist = append(vlist, defaultBranch)
 	}
 
 	return vlist, nil

--- a/gps/vcs_source_test.go
+++ b/gps/vcs_source_test.go
@@ -212,11 +212,11 @@ func testGopkginSourceInteractions(t *testing.T) {
 
 		vlist := hidePair(pvlist)
 		if len(vlist) != len(evl) {
-			t.Errorf("gopkgin test repo should've produced %v versions, got %v", len(evl), len(vlist))
+			t.Errorf("gopkgin test repo (%s) should've produced %v versions, got %v.\n%v", un, len(evl), len(vlist), vlist)
 		} else {
 			SortForUpgrade(vlist)
 			if !reflect.DeepEqual(vlist, evl) {
-				t.Errorf("Version list was not what we expected:\n\t(GOT): %s\n\t(WNT): %s", vlist, evl)
+				t.Errorf("Version list for %s was not what we expected:\n\t(GOT): %#v\n\t(WNT): %#v", un, vlist, evl)
 			}
 		}
 
@@ -232,7 +232,7 @@ func testGopkginSourceInteractions(t *testing.T) {
 		} else {
 			SortForUpgrade(vlist)
 			if !reflect.DeepEqual(vlist, evl) {
-				t.Errorf("Version list was not what we expected:\n\t(GOT): %s\n\t(WNT): %s", vlist, evl)
+				t.Errorf("Version list for %s was not what we expected:\n\t(GOT): %#v\n\t(WNT): %#v", un, vlist, evl)
 			}
 		}
 
@@ -247,7 +247,24 @@ func testGopkginSourceInteractions(t *testing.T) {
 
 	// simultaneously run for v1, v2, and v3 filters of the target repo
 	wg := &sync.WaitGroup{}
-	wg.Add(4)
+	wg.Add(6)
+
+	go func() {
+		// Treat master as v0 when no other branches/tags exist that match gopkg.in's rules
+		tfunc("gopkg.in/carolynvs/deptest-gopkgin-implicit-v0.v0", "github.com/carolynvs/deptest-gopkgin-implicit-v0", 0, []Version{
+			newDefaultBranch("notmaster").Pair(Revision("94ee631b9833cd805d15f50a52e0533124ec0292")),
+		})
+		wg.Done()
+	}()
+
+	go func() {
+		// Use the existing v0 branch for v0, not master
+		tfunc("gopkg.in/carolynvs/deptest-gopkgin-explicit-v0.v0", "github.com/carolynvs/deptest-gopkgin-explicit-v0", 0, []Version{
+			newDefaultBranch("v0").Pair(Revision("ec73e84554fb28f08dba630e48dbec868e77f734")),
+		})
+		wg.Done()
+	}()
+
 	go func() {
 		tfunc("gopkg.in/sdboyer/gpkt.v1", "github.com/sdboyer/gpkt", 1, []Version{
 			NewVersion("v1.1.0").Pair(Revision("b2cb48dda625f6640b34d9ffb664533359ac8b91")),

